File: go/pkg/encoders/string.go
package encoders

import (
	"errors"
	"unsafe"

	"github.com/splunk/stef/go/pkg"
)

type StringEncoder struct {
	buf     pkg.BytesWriter
	dict    *StringEncoderDict
	limiter *pkg.SizeLimiter
}

type StringEncoderDict struct {
	m       map[string]int
	limiter *pkg.SizeLimiter
}

func (e *StringEncoderDict) Init(limiter *pkg.SizeLimiter) {
	e.m = make(map[string]int)
	e.limiter = limiter
}

func (e *StringEncoderDict) Reset() {
	e.m = make(map[string]int)
}

func (e *StringEncoder) Init(dict *StringEncoderDict, limiter *pkg.SizeLimiter, columns *pkg.WriteColumnSet) error {
	e.dict = dict
	e.limiter = limiter
	return nil
}

func (e *StringEncoder) Encode(val string) {
	oldLen := len(e.buf.Bytes())
	if e.dict != nil {
		if refNum, exists := e.dict.m[val]; exists {
			e.buf.WriteVarint(int64(-refNum - 1))
			newLen := len(e.buf.Bytes())
			e.dict.limiter.AddFrameBytes(uint(newLen - oldLen))
			return
		}
	}
	strLen := len(val)
	if e.dict != nil && strLen > 1 {
		refNum := len(e.dict.m)
		e.dict.m[val] = refNum
		e.dict.limiter.AddDictElemSize(uint(strLen) + uint(unsafe.Sizeof(val)))
	}
	e.buf.WriteVarint(int64(strLen))
	e.buf.WriteStringBytes(val)
	newLen := len(e.buf.Bytes())
	e.limiter.AddFrameBytes(uint(newLen - oldLen))
}

func (e *StringEncoder) CollectColumns(columnSet *pkg.WriteColumnSet) {
	columnSet.SetBytes(&e.buf)
}

func (e *StringEncoder) WriteTo(buf *pkg.BytesWriter) {
	buf.WriteBytes(e.buf.Bytes())
}

func (e *StringEncoder) Reset() {
}

type StringDecoder struct {
	buf    pkg.BytesReader
	column *pkg.ReadableColumn
	dict   *StringDecoderDict
}

type StringDecoderDict struct {
	dict []string
}

func (d *StringDecoderDict) Init() {
}

var ErrInvalidRefNum = errors.New("invalid RefNum, out of dictionary range")

func (d *StringDecoder) Continue() {
	d.buf.Reset(d.column.Data())
}

func (d *StringDecoder) Reset() {
}

func (d *StringDecoder) Decode(dst *string) error {
	varint, err := d.buf.ReadVarint()
	if err != nil {
		return err
	}

	if varint >= 0 {
		strLen := int(varint)
		*dst, err = d.buf.ReadStringBytes(strLen)
		if err != nil {
			return err
		}
		if strLen > 1 && d.dict != nil {
			d.dict.dict = append(d.dict.dict, *dst)
		}
		return nil
	}
	if d.dict == nil {
		return ErrInvalidRefNum
	}
	refNum := int(-varint - 1)
	if refNum >= len(d.dict.dict) {
		return ErrInvalidRefNum
	}
	*dst = d.dict.dict[refNum]
	return nil
}

func (d *StringDecoder) Init(dict *StringDecoderDict, columns *pkg.ReadColumnSet) error {
	d.dict = dict
	d.column = columns.Column()
	return nil
}

File: rust/stef-core/src/codecs/string.rs
use crate::{errors::ERR_INVALID_REF_NUM, membuffer::{BytesReader, BytesWriter}, recordbuf::{ReadColumnSet, ReadableColumn, WriteColumnSet}, SizeLimiter};

/// Plain string encoder using length+bytes records.
#[derive(Default)]
pub struct StringEncoder {
    buf: BytesWriter,
    limiter: Option<*mut SizeLimiter>,
}

impl StringEncoder {
    pub fn init(&mut self, limiter: &mut SizeLimiter, _columns: &mut WriteColumnSet) {
        self.limiter = Some(limiter);
    }

    pub fn encode(&mut self, val: &str) {
        let old_len = self.buf.bytes().len();
        self.buf.write_varint(val.len() as i64);
        self.buf.write_string_bytes(val);
        let new_len = self.buf.bytes().len();
        if let Some(ptr) = self.limiter {
            unsafe { (*ptr).add_frame_bytes(new_len - old_len) };
        }
    }

    pub fn collect_columns(&mut self, column_set: &mut WriteColumnSet) {
        column_set.set_bytes(&mut self.buf);
    }

    pub fn reset(&mut self) {}
}

/// Plain string decoder.
#[derive(Default)]
pub struct StringDecoder {
    buf: BytesReader,
    column: Option<*mut ReadableColumn>,
}

impl StringDecoder {
    pub fn init(&mut self, columns: &mut ReadColumnSet) {
        self.column = Some(columns.column() as *mut ReadableColumn);
    }

    pub fn continue_(&mut self) {
        let column_ptr = self.column.expect("decoder not initialized");
        // Safe because generated code keeps read column tree alive for decoder lifetime.
        let data = unsafe { (&*column_ptr).data().to_vec() };
        self.buf.reset(data);
    }

    pub fn reset(&mut self) {}

    pub fn decode(&mut self, dst: &mut String) -> crate::errors::Result<()> {
        let varint = self.buf.read_varint()?;
        if varint >= 0 {
            *dst = self.buf.read_string_mapped(varint as usize)?;
            Ok(())
        } else {
            Err(ERR_INVALID_REF_NUM)
        }
    }
}
